Fix DTrace uint64_t pointers and introduce capabilitylen function

nicomazz opened the pull request:
To have a better description of the changes, take a look at each commit message.
A summary:

- DTrace and CTF use dynsym symbols
- Many uint64_t have been chanced to capabilities
- A new model has been added to CTF (to have sizeof pointer == 16)
- A new function has been added, able to get the len of a capability function parameter. This function was only a test to see what was needed to change.

nicomazz committed:
Fix CTFFLAGS in kern.post.mk

nicomazz committed:
Add new D language version

nicomazz committed:
Fix CTFFLAGS in kern.post.mk

Changing CTFFLAGS directly didn't work because it was used in other places that didn't require the -s flag

nicomazz committed:
Fix libdtrace symbols section for MIPS

In Mips we use .dynsym instead of .symtab for the kernel.
In addition, the kernel linker file doesn't have the correct path in mips, so I set it manually.

nicomazz committed:
Add new CTF and DOF Model

A new model with the correct pointer size (16) has been added to both CTF and DOF (DTrace object format)

nicomazz committed:
Fix kernel module path replacement in MIPS

Now it is more future proof

nicomazz committed:
Add capabilitylen function to DTrace

This is only a test function that returns the length of a capability argument.
it is possible to test it with:
`dtrace -S -x nolibs -n 'fbt::user_ioctl:entry { printf("%d",capabilitylen(arg3)); }'`

The function fbt_getargdesc is not static anymore, because it is needed from fbt_isa.c to get the parameters.

nicomazz committed:
Change DTrace uint64 to capabilities

While it (should) work even with a noncap userspace, it will never with 32bits.
However, there are multiple reasons why it didn't work with 32 bits not even before.

The `stack()` function has been temporarily commented out (I will fix it, I promise)

Places where int64_t has been changed:
- dtrace_probe()
- dof interpreter (registers are now all uintcap_t). A few more fixes will be needed here.
- dtrace_key_t
- dtrace_mstate_t args
- sdt_probe_stub

nicomazz committed:
Fix precompiler flag for cheri CTF model

nicomazz committed:
Merge branch 'dtrace_support' of github.com:nicomazz/cheribsd into dtrace_cap_parameters

nicomazz committed:
Merge branch 'dtrace_support' of github.com:nicomazz/cheribsd into dtrace_cap_parameters

nicomazz committed:
Fix shift in D compiler for capabilities

Somewhere, the dt_node_t for the arg array (args[]) is set to contain int64_t (size == 8). Now it contains capabilities, so that code should be changed accordingly.
This is not a definitive fix, but so far I haven't seen a case where it fails. The tests should catch them.

qwattash commented on cddl/contrib/opensolaris/lib/libdtrace/common/dt_open.c:
```diff
@@ -220,6 +222,9 @@ static const dt_ident_t _dtrace_globals[] = {
 { "breakpoint", DT_IDENT_ACTFUNC, 0, DT_ACT_BREAKPOINT,
 	DT_ATTR_STABCMN, DT_VERS_1_0,
 	&dt_idops_func, "void()" },
+{ "capabilitylen", DT_IDENT_FUNC, 0, DIF_SUBR_CAPABILITY_LEN,
```
Should only be available when we have `__has_feature(capabilities)`?

qwattash commented on sys/cddl/contrib/opensolaris/uts/common/dtrace/dtrace.c:
```diff
@@ -3270,7 +3278,7 @@ dtrace_dif_varstr(uintptr_t addr, dtrace_state_t *state,
  * builtin values such as 'execargs'.
  */
-static uintptr_t
+static uintcap_t
```
Note that the local variable `ret` here is still an `uintptr_t` so the return value will not be a valid capability.

qwattash commented on sys/cddl/contrib/opensolaris/uts/common/dtrace/dtrace.c:
```diff
@@ -3224,7 +3232,7 @@ dtrace_speculation_buffer(dtrace_state_t *state, processorid_t cpuid,
  * builtin values such as 'execname' and 'probefunc.'
  */
-uintptr_t
+uintcap_t
```
Note that the local variable `ret` here is still an `uintptr_t` so the return value will not be a valid capability.

dstolfa commented on cddl/contrib/opensolaris/lib/libdtrace/common/dt_module.c:
```diff
@@ -1179,6 +1200,10 @@ dt_module_update(dtrace_hdl_t *dtp, struct kld_file_stat *k_stat)
 	(void) strlcpy(fname, k_stat->pathname, sizeof(fname));
+#ifdef __mips__
+	if (strncmp(fname, "kernel", sizeof(name)) == 0)
+		strlcpy(fname,"/boot/kernel/kernel",sizeof(fname));
```
If I understand this correctly, it's trying to simply get the path to the currently running kernel, or the one specified by some variable?

Consider using the `kern.bootfile` sysctl if so.

nicomazz replied:
It's trying to get the path of the current running kernel.
It seems that in MIPS the filename remains to `kernel`, and dtrace wasn't able to read the CTF.
I'll use kern.bootfile, thanks!

nicomazz replied:
fixed and moved to https://github.com/CTSRD-CHERI/cheribsd/pull/466/files

dstolfa commented on cddl/contrib/opensolaris/lib/libdtrace/common/dt_open.c:
```diff
@@ -159,6 +160,7 @@ const dt_version_t _dtrace_versions[] = {
 	DT_VERS_1_12,	/* D API 1.12 */
 	DT_VERS_1_12_1,	/* D API 1.12.1 */
 	DT_VERS_1_13,	/* D API 1.13 */
+	DT_VERS_1_13_1,	/* D API 1.13.1 Capabilities support */
```
This isn't really a minor version bump, as it breaks some semantics that D/DIF assume. I would bump a major version, possibly making it `DT_VERS_2_0`.

dstolfa commented on sys/cddl/contrib/opensolaris/uts/common/dtrace/dtrace.c:
```diff
@@ -6151,10 +6165,10 @@ dtrace_dif_emulate(dtrace_difo_t *difo, dtrace_mstate_t *mstate,
 			break;
 		case DIF_OP_ADD:
-			regs[rd] = regs[r1] + regs[r2];
+			regs[rd] = (uint64_t)regs[r1] + regs[r2];
```
This line breaks the most of D scripts by making it impossible to retrieve a capability from it. The DTrace compiler will emit the following sequence for any kind of pointer following:

```
syscall::read:entry { x = curthread->td_tid; }

OFF OPCODE      INSTRUCTION
00: 29010001    ldgs DT_VAR(256), %r1		! DT_VAR(256) = "curthread"
01: 25000002    setx DT_INTEGER[0], %r2		! 0xa0
02: 07010201    add  %r1, %r2, %r1
03: 1e010001    ldsw [%r1], %r1
04: 23000001    ret  %r1
```

Notice how we have a `ldgs DT_VAR(256), %r1` which loads the value of the `curthread` variable into %r1, making it of type `struct thread *`. Later on, we have `add  %r1, %r2, %r1`, which adds a `uint64_t` to `%r1` in order to reach the `td_tid` struct member, followed by a `ldsw [%r1], %r1`, which dereferences whatever is in `%r1` and stores the result in `%r1`. Since this casts the capability to an `uint64_t`, this can never be dereferenced in an `ldub/lduh/lduw/ldsb/ldsh/ldsw/ldx` instruction, since it is an invalid capability. The same might be true for some other instructions.

nicomazz replied:
True, I think this is a left out from a previous change. Now `regs` contains `uintcap_t`, sho the cast should not be needed anymore

dstolfa commented on sys/cddl/contrib/opensolaris/uts/common/dtrace/dtrace.c:
```diff
@@ -6097,15 +6110,15 @@ dtrace_dif_emulate(dtrace_difo_t *difo, dtrace_mstate_t *mstate,
 
 	dtrace_key_t tupregs[DIF_DTR_NREGS + 2]; /* +2 for thread and id */
-	uint64_t regs[DIF_DIR_NREGS];
+	uintcap_t regs[DIF_DIR_NREGS];
```
There is another, extremely subtle thing that happens with registers and the `setx` instruction. I was unsure where to comment on it, so I will put it here.

```
dtrace -S -e -n 'syscall::read:entry { printf("%p\n", &`prison0); }'

DIFO 0x8009f0070 returns D type (pointer) (size 8)
OFF OPCODE      INSTRUCTION
00: 25000001    setx DT_INTEGER[0], %r1		! 0xffffffff81b0c7f0
01: 23000001    ret  %r1
```

Notice how ```&`prison0``` is a **pointer**, but is being created in the **integer table** as an integer. There is absolutely zero distinction between pointers and integers in the DTrace abstract machine, and this has to change in order for `setx` to work correctly.

dstolfa commented on sys/cddl/contrib/opensolaris/uts/common/sys/dtrace.h:
```diff
@@ -315,7 +315,8 @@ typedef enum dtrace_probespec {
 #define	DIF_SUBR_JSON			52
 #define	DIF_SUBR_STRTOLL		53
-#define	DIF_SUBR_MAX			53	/* max subroutine value */
+#define DIF_SUBR_CAPABILITY_LEN		54
```
Tab instead of space :-)

nicomazz committed:
Merge branch 'dtrace_support' into dtrace_cap_parameters

nicomazz committed:
Read bootfile from ioctl

nicomazz committed:
Enable capability_len only for cheri and bump version D version number

nicomazz committed:
Revert a few useless uintcap_t

nicomazz committed:
Add checks for capabilities in dtrace_dif_emulate

Now, the sum case in handled using the valid operand as LHS, and we assert that both operads have tag cleared for other operations.

nicomazz committed:
Merge branch 'dtrace_support' into dtrace_cap_parameters

nicomazz committed:
Add capability related functions to DTrace
